tests/python_package_test/test_interaction_constraints.py

```diff
@@ -0,0 +1,134 @@
+import numpy as np
+import pandas as pd
+import pytest
+
+import lightgbm as lgb
+
+
+def simple_pd(X, estimator, feature, values):
+    """Calculate simple partial dependency."""
+    Xc = X.copy()
+    yps = np.zeros_like(values)
+    for i, x in enumerate(values):
+        Xc[feature] = x
+        yps[i] = np.mean(estimator.predict(Xc.values))
+    return values, yps
+
+
+def simple_pds(df, estimator, features):
+    """Calculate simple partial dependency for all features."""
+    pds = {}
+    for feat in features:
+        values, yps = simple_pd(
+            df[features], estimator, feat, np.sort(df[feat].unique())
+        )
+        pds[feat] = pd.DataFrame(data={feat: values, "y": yps})
+    return pds
+
+
+@pytest.fixture
+def make_data():
+    """Make toy data."""
+    np.random.seed(1)
+    n = 10000
+    d = 3
+    X = np.random.normal(size=(n, d))
+    # round to speed things up
+    X = np.round(X)
+    eps = np.random.normal() * 0.1
+    y = -1 * X[:, 0] + 3 * X[:, 1] + X[:, 0] * X[:, 1] + np.cos(X[:, 2]) + eps
+    df = pd.DataFrame(data={"y": y, "x0": X[:, 0], "x1": X[:, 1], "x2": X[:, 2]})
+    features = ["x0", "x1", "x2"]
+    outcome = "y"
+
+    return df, features, outcome
+
+
+def find_interactions(gbm, feature_sets):
+    """Find interactions in tree.
+
+    Parameters
+    ---------
+    gbm: booster
+    feature_sets: list of list
+        set of features across which to check for interactions
+
+    Returns
+    -------
+    tree_features: pandas.DataFrame
+        boolean flag for every tree of has interaction across feature sets.
+    """
+    df_trees = gbm.trees_to_dataframe()
+    tree_features = (
+        df_trees.groupby("tree_index")
+        .apply(lambda x: set(x["split_feature"]) - set([None]))
+        .reset_index()
+        .rename(columns={0: "features"})
+    )
+
+    def has_interaction(tree):
+        n = 0
+        for fs in feature_sets:
+            if len(tree["features"].intersection(fs)) > 0:
+                n += 1
+        if n > 1:
+            return True
+        else:
+            return False
+
+    tree_features["has_interaction"] = tree_features.apply(has_interaction, axis=1)
+
+    return tree_features
+
+
+@pytest.fixture
+def get_boosting_params():
+    boosting_params = {
+        "boosting_type": "gbdt",
+        "objective": "regression",
+        "num_leaves": 5,
+        "learning_rate": 0.1,
+        "num_boost_round": 100,
+        # disallow all interactions
+        "interaction_constraints": [[0], [1], [2]],
+        "monotone_constraints": [1, 1, 0],
+        "monotone_constraints_method": "basic",
+    }
+    return boosting_params
+
+
+@pytest.mark.parametrize(
+    "monotone_constraints_method", ["basic", "intermediate", "advanced"]
+)
+def test_interaction_constraints(
```

Comment:
Please merge this new test into already existent one:
https://github.com/microsoft/LightGBM/blob/37e987828d3141aded1a6e356fdcc7954a6f2f5c/tests/python_package_test/test_engine.py#L1255

For the reference: here is `test_interaction_constraints`
https://github.com/microsoft/LightGBM/blob/37e987828d3141aded1a6e356fdcc7954a6f2f5c/tests/python_package_test/test_engine.py#L2465

Reply:
Thanks for the comments! Just to clarify, you asked that I add the test to `test_monotone_constraints()` or `test_interaction_constraints()`? Seems to me that `test_interaction_constraints()` would be the better place for it. Alternatively, I could just add a new function `test_monotone_and_interaction_constraints()` to  `test_engine.py` that does what I'm currently doing in `test_interaction_constraints.py`.

Reply:
ok. had closer look at the tests. I think I can merge into `test_monotone_constraints()`.

Reply:
I think any variant will be fine, but better will be one that reuses existing test infrastructure as much as possible. Thank you!

Reply:
I updated `test_monotone_constraints()` as discussed.

Reply:
Thanks a lot!